File: src/modules/audit/audit.controller.ts
import {
  Controller,
  Get,
  Post,
  Body,
  Param,
  Query,
  UseGuards,
  Request,
} from '@nestjs/common';
import { ApiTags, ApiOperation, ApiResponse, ApiBearerAuth, ApiQuery } from '@nestjs/swagger';
import { AuditService } from './audit.service';
import { CreateAuditLogDto } from './dto/create-audit-log.dto';
import { JwtAuthGuard } from '../auth/guards/jwt-auth.guard';
import { RolesGuard } from '../auth/guards/roles.guard';
import { Roles } from '../auth/decorators/roles.decorator';

@ApiTags('Auditoría')
@Controller('audit')
@UseGuards(JwtAuthGuard, RolesGuard)
@ApiBearerAuth()
export class AuditController {
  constructor(private readonly auditService: AuditService) {}

  @Post()
  @ApiOperation({ summary: 'Crear log de auditoría' })
  @ApiResponse({ status: 201, description: 'Log de auditoría creado exitosamente' })
  @ApiResponse({ status: 400, description: 'Datos inválidos' })
  @Roles('admin')
  async create(@Body() createAuditLogDto: CreateAuditLogDto, @Request() req) {
    return this.auditService.create(createAuditLogDto, req.user.id);
  }

  @Get()
  @ApiOperation({ summary: 'Obtener logs de auditoría con filtros' })
  @ApiResponse({ status: 200, description: 'Lista de logs de auditoría' })
  @ApiQuery({ name: 'page', required: false, type: Number })
  @ApiQuery({ name: 'limit', required: false, type: Number })
  @ApiQuery({ name: 'userId', required: false, type: Number })
  @ApiQuery({ name: 'action', required: false, type: String })
  @ApiQuery({ name: 'resource', required: false, type: String })
  @ApiQuery({ name: 'startDate', required: false, type: String })
  @ApiQuery({ name: 'endDate', required: false, type: String })
  @ApiQuery({ name: 'search', required: false, type: String })
  @Roles('admin', 'supervisor')
  async findAll(
    @Query('page') page: string = '1',
    @Query('limit') limit: string = '50',
    @Query('userId') userId?: string,
    @Query('action') action?: string,
    @Query('resource') resource?: string,
    @Query('startDate') startDate?: string,
    @Query('endDate') endDate?: string,
    @Query('search') search?: string,
  ) {
    const filters = {
      userId: userId ? parseInt(userId) : undefined,
      action,
      resource,
      startDate: startDate ? new Date(startDate) : undefined,
      endDate: endDate ? new Date(endDate) : undefined,
      search,
    };

    return this.auditService.findAll(parseInt(page), parseInt(limit), filters);
  }

  @Get('stats')
  @ApiOperation({ summary: 'Obtener estadísticas de auditoría' })
  @ApiResponse({ status: 200, description: 'Estadísticas de auditoría' })
  @ApiQuery({ name: 'days', required: false, type: Number, description: 'Días hacia atrás (default: 30)' })
  @Roles('admin')
  async getStats(@Query('days') days: string = '30') {
    return this.auditService.getStats(parseInt(days));
  }

  @Get('recent')
  @ApiOperation({ summary: 'Obtener actividad reciente' })
  @ApiResponse({ status: 200, description: 'Actividad reciente' })
  @ApiQuery({ name: 'limit', required: false, type: Number })
  @Roles('admin', 'supervisor')
  async getRecentActivity(@Query('limit') limit: string = '20') {
    return this.auditService.getRecentActivity(parseInt(limit));
  }

  @Get('user/:userId')
  @ApiOperation({ summary: 'Obtener logs de un usuario específico' })
  @ApiResponse({ status: 200, description: 'Logs del usuario' })
  @ApiQuery({ name: 'limit', required: false, type: Number })
  @Roles('admin')
  async findByUser(
    @Param('userId') userId: string,
    @Query('limit') limit: string = '100',
  ) {
    return this.auditService.findByUser(parseInt(userId), parseInt(limit));
  }

  @Get('action/:action')
  @ApiOperation({ summary: 'Obtener logs por acción específica' })
  @ApiResponse({ status: 200, description: 'Logs de la acción' })
  @ApiQuery({ name: 'limit', required: false, type: Number })
  @Roles('admin')
  async findByAction(
    @Param('action') action: string,
    @Query('limit') limit: string = '100',
  ) {
    return this.auditService.findByAction(action, parseInt(limit));
  }

  @Get('resource/:resource')
  @ApiOperation({ summary: 'Obtener logs por recurso específico' })
  @ApiResponse({ status: 200, description: 'Logs del recurso' })
  @ApiQuery({ name: 'limit', required: false, type: Number })
  @Roles('admin')
  async findByResource(
    @Param('resource') resource: string,
    @Query('limit') limit: string = '100',
  ) {
    return this.auditService.findByResource(resource, parseInt(limit));
  }

  @Get(':id')
  @ApiOperation({ summary: 'Obtener log específico por ID' })
  @ApiResponse({ status: 200, description: 'Log encontrado' })
  @ApiResponse({ status: 404, description: 'Log no encontrado' })
  @Roles('admin')
  async findOne(@Param('id') id: string) {
    return this.auditService.findOne(parseInt(id));
  }
} 

File: src/main/java/com/yego/backend/controller/yego-principal/AuditController.java
package com.yego.backend.controller.yego_principal;

import com.yego.backend.entity.yego_principal.api.*;
import com.yego.backend.service.yego_principal.AuditService;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.security.access.prepost.PreAuthorize;
import org.springframework.security.core.Authentication;
import org.springframework.web.bind.annotation.*;

import jakarta.persistence.EntityNotFoundException;
import jakarta.validation.Valid;
import java.time.LocalDate;
import java.time.LocalDateTime;
import java.time.format.DateTimeFormatter;
import java.util.List;
import java.util.Map;

/**
 * Controlador REST para auditoría del sistema YEGO Principal
 * Equivalente a AuditController de NestJS
 */
@Slf4j
@RestController
@RequestMapping("/api/yego-principal/audit")
@RequiredArgsConstructor
public class AuditController {
    
    private final AuditService auditService;
    
    /**
     * Crear log de auditoría
     */
    @PostMapping
    @PreAuthorize("hasRole('ADMIN')")
    public ResponseEntity<?> create(@Valid @RequestBody CreateAuditLogDto createAuditLogDto,
                                   Authentication authentication) {
        try {
            Long userId = Long.parseLong(authentication.getName());
            AuditLogResponseDto auditLog = auditService.create(createAuditLogDto, userId);
            
            return ResponseEntity.status(HttpStatus.CREATED).body(auditLog);
            
        } catch (Exception e) {
            log.error("Error creando log de auditoría YEGO Principal: {}", e.getMessage());
            return ResponseEntity.status(HttpStatus.INTERNAL_SERVER_ERROR)
                    .body(Map.of("message", "Error interno del servidor"));
        }
    }
    
    /**
     * Obtener logs de auditoría con filtros
     */
    @GetMapping
    @PreAuthorize("hasRole('ADMIN') or hasRole('SUPERVISOR')")
    public ResponseEntity<?> findAll(
            @RequestParam(defaultValue = "1") Integer page,
            @RequestParam(defaultValue = "50") Integer limit,
            @RequestParam(required = false) Long userId,
            @RequestParam(required = false) String action,
            @RequestParam(required = false) String resource,
            @RequestParam(required = false) String startDate,
            @RequestParam(required = false) String endDate,
            @RequestParam(required = false) String search) {
        
        try {
            AuditFilterDto filters = AuditFilterDto.builder()
                    .userId(userId)
                    .action(action)
                    .resource(resource)
                    .startDate(startDate != null ? LocalDate.parse(startDate).atStartOfDay() : null)
                    .endDate(endDate != null ? LocalDate.parse(endDate).atTime(23, 59, 59) : null)
                    .search(search)
                    .build();
            
            AuditLogPageDto result = auditService.findAll(page, limit, filters);
            
            return ResponseEntity.ok(result);
            
        } catch (Exception e) {
            log.error("Error obteniendo logs de auditoría YEGO Principal: {}", e.getMessage());
            return ResponseEntity.status(HttpStatus.INTERNAL_SERVER_ERROR)
                    .body(Map.of("message", "Error interno del servidor"));
        }
    }
    
    /**
     * Obtener estadísticas de auditoría
     */
    @GetMapping("/stats")
    @PreAuthorize("hasRole('ADMIN')")
    public ResponseEntity<?> getStats(@RequestParam(defaultValue = "30") Integer days) {
        try {
            AuditStatsDto stats = auditService.getStats(days);
            return ResponseEntity.ok(stats);
            
        } catch (Exception e) {
            log.error("Error obteniendo estadísticas de auditoría YEGO Principal: {}", e.getMessage());
            return ResponseEntity.status(HttpStatus.INTERNAL_SERVER_ERROR)
                    .body(Map.of("message", "Error interno del servidor"));
        }
    }
    
    /**
     * Obtener actividad reciente
     */
    @GetMapping("/recent")
    @PreAuthorize("hasRole('ADMIN') or hasRole('SUPERVISOR')")
    public ResponseEntity<?> getRecentActivity(@RequestParam(defaultValue = "20") Integer limit) {
        try {
            List<AuditLogResponseDto> recentActivity = auditService.getRecentActivity(limit);
            return ResponseEntity.ok(recentActivity);
            
        } catch (Exception e) {
            log.error("Error obteniendo actividad reciente YEGO Principal: {}", e.getMessage());
            return ResponseEntity.status(HttpStatus.INTERNAL_SERVER_ERROR)
                    .body(Map.of("message", "Error interno del servidor"));
        }
    }
    
    /**
     * Obtener logs de un usuario específico
     */
    @GetMapping("/user/{userId}")
    @PreAuthorize("hasRole('ADMIN')")
    public ResponseEntity<?> findByUser(@PathVariable Long userId,
                                       @RequestParam(defaultValue = "100") Integer limit) {
        try {
            List<AuditLogResponseDto> userLogs = auditService.findByUser(userId, limit);
            return ResponseEntity.ok(userLogs);
            
        } catch (Exception e) {
            log.error("Error obteniendo logs de usuario YEGO Principal {}: {}", userId, e.getMessage());
            return ResponseEntity.status(HttpStatus.INTERNAL_SERVER_ERROR)
                    .body(Map.of("message", "Error interno del servidor"));
        }
    }
    
    /**
     * Obtener logs por acción específica
     */
    @GetMapping("/action/{action}")
    @PreAuthorize("hasRole('ADMIN')")
    public ResponseEntity<?> findByAction(@PathVariable String action,
                                         @RequestParam(defaultValue = "100") Integer limit) {
        try {
            List<AuditLogResponseDto> actionLogs = auditService.findByAction(action, limit);
            return ResponseEntity.ok(actionLogs);
            
        } catch (Exception e) {
            log.error("Error obteniendo logs por acción YEGO Principal {}: {}", action, e.getMessage());
            return ResponseEntity.status(HttpStatus.INTERNAL_SERVER_ERROR)
                    .body(Map.of("message", "Error interno del servidor"));
        }
    }
    
    /**
     * Obtener logs por recurso específico
     */
    @GetMapping("/resource/{resource}")
    @PreAuthorize("hasRole('ADMIN')")
    public ResponseEntity<?> findByResource(@PathVariable String resource,
                                           @RequestParam(defaultValue = "100") Integer limit) {
        try {
            List<AuditLogResponseDto> resourceLogs = auditService.findByResource(resource, limit);
            return ResponseEntity.ok(resourceLogs);
            
        } catch (Exception e) {
            log.error("Error obteniendo logs por recurso YEGO Principal {}: {}", resource, e.getMessage());
            return ResponseEntity.status(HttpStatus.INTERNAL_SERVER_ERROR)
                    .body(Map.of("message", "Error interno del servidor"));
        }
    }
    
    /**
     * Obtener log específico por ID
     */
    @GetMapping("/{id}")
    @PreAuthorize("hasRole('ADMIN')")
    public ResponseEntity<?> findOne(@PathVariable Long id) {
        try {
            AuditLogResponseDto auditLog = auditService.findOne(id);
            return ResponseEntity.ok(auditLog);
            
        } catch (EntityNotFoundException e) {
            return ResponseEntity.status(HttpStatus.NOT_FOUND)
                    .body(Map.of("message", e.getMessage()));
        } catch (Exception e) {
            log.error("Error obteniendo log de auditoría YEGO Principal {}: {}", id, e.getMessage());
            return ResponseEntity.status(HttpStatus.INTERNAL_SERVER_ERROR)
                    .body(Map.of("message", "Error interno del servidor"));
        }
    }
}
